reps/2023-03-10-batch-inference.md

```diff
@@ -0,0 +1,143 @@
+
+```python
+def map_predictor(create_predictor: Callable[[], Predictor]], 
+                  batch_size: Optional[int], 
+                  feature_columns: Optional[List[str]] = None, 
+                  keep_columns: Optional[List[str]] = None, 
+                  prefetch_batches: Optional[int] = 0, 
+                  min_scoring_workers: int = 1,
+                  max_scoring_workers: Optional[int] = None,
+                  num_cpus_per_worker: Optional[int] = None,
+                  num_gpus_per_worker: Optional[int] = None, 
+                  **predict_kwargs):
+  """Predict on batches of data.
+
+  Args:
+    predictor: A function that returns a Predictor to use for inference.
+    batch_size: Batch size to use for prediction.
+    feature_columns: List of columns in the preprocessed dataset to use for
+      prediction. Columns not specified will be dropped
+      from `data` before being passed to the predictor.
+      If None, use all columns in the preprocessed dataset.
+    keep_columns: List of columns in the preprocessed dataset to include
+        in the prediction result. This is useful for calculating final
+        accuracies/metrics on the result dataset. If None,
+        the columns in the output dataset will contain
+        just the prediction results.
+    min_scoring_workers: Minimum number of scoring actors.
+    max_scoring_workers: If set, specify the maximum number of scoring actors.
+    num_cpus_per_worker: Number of CPUs to allocate per scoring worker.
+      Set to 1 by default.
+    num_gpus_per_worker: Number of GPUs to allocate per scoring worker.
+      Set to 0 by default. If you want to use GPUs for inference, please
+      specify this parameter.
+    ray_remote_args: Additional resource requirements to request from ray.
+    predict_kwargs: Keyword arguments passed to the predictor's
+        ``predict()`` method.
+  """
+```
+3. Deprecate `BatchMapper`, as the same functionality can be achieved via `map_batches`.
+      
+## Example: Multi-stage Inference Pipeline from pre-trained model
+
+Note that Checkpoints are not on the critical path, but can still be used.
+
+### Before
+```python
+preprocessor1 = BatchMapper(lambda x: x+1)
+preprocessor2 = TorchVisionPreprocessor(torchvision.transforms.Crop(224, 224))
+chained = Chain([preprocessor1, preprocessor2])
+
+model = pretrained_resnet()
+checkpoint = TorchCheckpoint.from_model(model, preprocessor=chained)
+batch_predictor = BatchPredictor.from_checkpoint(checkpoint)
+
+ds = ds.read_parquet("s3://data")
+predictions = batch_predictor.predict(ds)
+```
+
+### After
+```python
+predictor = lambda: TorchPredictor(pretrained_resnet())
+
+# Alternatively, create from AIR Checkpoint.
+# predictor = lambda: TorchPredictor.from_checkpoint(air_checkpoint)
+
+predictions = ds.read_parquet("s3://data")
+    .map_batches(lambda x: x+1)
+    .map_preprocessor(TorchVisionPreprocessor(torchvision.transforms.Crop(224, 224)))
```

Comment:
like the new dataset-centric way of doing things.
a stupid question, why do we have to differentiate preprocessor vs predictor?
can they all simply use the map_batches() api?
```
predictions = ds.read_parquet("s3://data")
    .map_batches(lambda x: x+1)
    .map_batches(TorchVisionPreprocessor(torchvision.transforms.Crop(224, 224)))
    .map_batches(TorchPredictor(pretrained_resnet()))
```
?

Reply:
1. Avoid overloading the existing `map_batches` API and make the API more explicit
2. Not all of the `map_batches` arguments are compatible or needed for preprocessors/predictors.